18MS: Implement private P1 and P2 + general tile lay anywhere

perwestling opened the pull request:
This PR implements private P1 and P2 in 18MS, which are privates that can lay extra yellow tiles during tile lay.
This require the possibility to specify empty hexes and tiles for the ability, so supporting this in the general code has been added.

tobymao commented on lib/engine/step/special_track.rb:
```diff
@@ -35,17 +35,18 @@ def process_lay_tile(action)
-          .tiles
-          .map { |name| @game.tiles.find { |t| t.name == name } }
+        tiles = tile_ability.tiles.map { |name| @game.tiles.find { |t| t.name == name } }
+        tiles = @game.tiles.uniq(&:name) if tiles.empty?
```
this seems like it could be bug prone. what happens if you've run out of tiles, then you allow all?

instead shouldn't you check if tile_abilities doesn't have any tiles?

perwestling replied:
Check tile_ability.tiles instead.

tobymao commented on lib/engine/step/tracker.rb:
```diff
@@ -71,14 +71,22 @@ def lay_tile(action, extra_cost: 0, entity: nil)
+
+          corporation = entity.owner
+          @game.game_error("Track laid in must be connected to one of #{corporation.id}'s stations") if
+            !@game.graph.reachable_hexes(corporation)[hex] && hex.name != corporation.coordinates
```
will this break 1889? is this rule actually generic, also you shouldn't do reachable checks outside of a game.loading for performance reasons

perwestling replied:
The only tile_lay ability outside 18MS with no hexes specified is NWR in 1882. The 1889 one have hexes as it specifies the coast towns.
Now I suppose 1882 should work anyway, but I have added 
  ability.when == 'track' &&
which excludes that ability.

But does not 1882 NWR require the tile lay to be connected?

An alternative is to add an attribute that specifies if tile lay need to be connected and use that to decide if this check should be done or not.

tobymao commented on lib/engine/step/tracker.rb:
```diff
@@ -71,14 +71,22 @@ def lay_tile(action, extra_cost: 0, entity: nil)
 
         entity.abilities(:teleport) do |ability, _|
-          ability.use! if ability.hexes.include?(hex.id) && ability.tiles.include?(tile.name)
+          ability.use! if ability.hexes.any? && ability.hexes.include?(hex.id) && ability.tiles.include?(tile.name)
```
don't you want to use the ability regardless? if abilitiy.hexes is an empty array, you'll never use the teleport ability

perwestling replied:
Changed backed.

tobymao commented on lib/engine/step/tracker.rb:
```diff
@@ -71,14 +71,22 @@ def lay_tile(action, extra_cost: 0, entity: nil)
+        end if entity.company?
+
+        entity.abilities(:tile_lay) do |ability|
+          next if ability.hexes.any? && (!ability.hexes.include?(hex.id) || !ability.tiles.include?(tile.name))
```
same here

perwestling replied:
I think it is correct, after the change of the code. We do want to use free and discount ability attributes if hexes is empty. 
In case of the tile_lay in 1882 it will mean that the following code is exercised, which is wasn't before but the NWR does not specify free or discount, so it will use default which is free=false and discount=0, so it wont change anything.

So this should be OK.

perwestling commented:
@tobymao Have now updated based on your comments, and rebased the commits.
The problem of doing tile lay outside of track lay still remains but that is a known problem.

tobymao commented on lib/engine/step/tracker.rb:
```diff
@@ -72,7 +72,15 @@ def lay_tile(action, extra_cost: 0, entity: nil, spender: nil)
+          if ability.reachable &&
+            !@game.loading &&
+            !@game.graph.reachable_hexes(spender)[hex] &&
+            hex.name != spender.coordinates
```
move the hex.name check first to improve performance, also can we make it inline

game_error if x &&
y &&
z

perwestling replied:
Correction pushed.
